// cloudflare@3.5.0/resources/page-shield/page-shield.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../resource.mjs";
import * as ConnectionsAPI from "./connections.mjs";
import * as CookiesAPI from "./cookies.mjs";
import * as PoliciesAPI from "./policies.mjs";
import * as ScriptsAPI from "./scripts.mjs";
export class PageShield extends APIResource {
    constructor() {
        super(...arguments);
        this.policies = new PoliciesAPI.Policies(this._client);
        this.connections = new ConnectionsAPI.Connections(this._client);
        this.scripts = new ScriptsAPI.Scripts(this._client);
        this.cookies = new CookiesAPI.Cookies(this._client);
    }
    /**
     * Updates Page Shield settings.
     */
    update(params, options) {
        const { zone_id, ...body } = params;
        return this._client.put(`/zones/${zone_id}/page_shield`, { body, ...options })._thenUnwrap((obj) => obj.result);
    }
    /**
     * Fetches the Page Shield settings.
     */
    get(params, options) {
        const { zone_id } = params;
        return this._client.get(`/zones/${zone_id}/page_shield`, options)._thenUnwrap((obj) => obj.result);
    }
}
(function (PageShield) {
    PageShield.Policies = PoliciesAPI.Policies;
    PageShield.PolicyListResponsesSinglePage = PoliciesAPI.PolicyListResponsesSinglePage;
    PageShield.Connections = ConnectionsAPI.Connections;
    PageShield.ConnectionsSinglePage = ConnectionsAPI.ConnectionsSinglePage;
    PageShield.Scripts = ScriptsAPI.Scripts;
    PageShield.ScriptsSinglePage = ScriptsAPI.ScriptsSinglePage;
    PageShield.Cookies = CookiesAPI.Cookies;
    PageShield.CookieListResponsesSinglePage = CookiesAPI.CookieListResponsesSinglePage;
})(PageShield || (PageShield = {}));
