// cloudflare@3.5.0/resources/api-gateway/api-gateway.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../resource.mjs";
import * as ConfigurationsAPI from "./configurations.mjs";
import * as SchemasAPI from "./schemas.mjs";
import * as DiscoveryAPI from "./discovery/discovery.mjs";
import * as OperationsAPI from "./operations/operations.mjs";
import * as SettingsAPI from "./settings/settings.mjs";
import * as UserSchemasAPI from "./user-schemas/user-schemas.mjs";
export class APIGateway extends APIResource {
    constructor() {
        super(...arguments);
        this.configurations = new ConfigurationsAPI.Configurations(this._client);
        this.discovery = new DiscoveryAPI.Discovery(this._client);
        this.operations = new OperationsAPI.Operations(this._client);
        this.schemas = new SchemasAPI.Schemas(this._client);
        this.settings = new SettingsAPI.Settings(this._client);
        this.userSchemas = new UserSchemasAPI.UserSchemas(this._client);
    }
}
(function (APIGateway) {
    APIGateway.Configurations = ConfigurationsAPI.Configurations;
    APIGateway.Discovery = DiscoveryAPI.Discovery;
    APIGateway.Operations = OperationsAPI.Operations;
    APIGateway.OperationListResponsesV4PagePaginationArray = OperationsAPI.OperationListResponsesV4PagePaginationArray;
    APIGateway.Schemas = SchemasAPI.Schemas;
    APIGateway.Settings = SettingsAPI.Settings;
    APIGateway.UserSchemas = UserSchemasAPI.UserSchemas;
    APIGateway.PublicSchemasV4PagePaginationArray = UserSchemasAPI.PublicSchemasV4PagePaginationArray;
})(APIGateway || (APIGateway = {}));
