// cloudflare@3.5.0/resources/magic-transit/magic-transit.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../resource.mjs";
import * as AppsAPI from "./apps.mjs";
import * as CfInterconnectsAPI from "./cf-interconnects.mjs";
import * as ConnectorsAPI from "./connectors.mjs";
import * as GRETunnelsAPI from "./gre-tunnels.mjs";
import * as IPSECTunnelsAPI from "./ipsec-tunnels.mjs";
import * as RoutesAPI from "./routes.mjs";
import * as SitesAPI from "./sites/sites.mjs";
export class MagicTransit extends APIResource {
    constructor() {
        super(...arguments);
        this.apps = new AppsAPI.Apps(this._client);
        this.cfInterconnects = new CfInterconnectsAPI.CfInterconnects(this._client);
        this.greTunnels = new GRETunnelsAPI.GRETunnels(this._client);
        this.ipsecTunnels = new IPSECTunnelsAPI.IPSECTunnels(this._client);
        this.routes = new RoutesAPI.Routes(this._client);
        this.sites = new SitesAPI.Sites(this._client);
        this.connectors = new ConnectorsAPI.Connectors(this._client);
    }
}
(function (MagicTransit) {
    MagicTransit.Apps = AppsAPI.Apps;
    MagicTransit.AppListResponsesSinglePage = AppsAPI.AppListResponsesSinglePage;
    MagicTransit.CfInterconnects = CfInterconnectsAPI.CfInterconnects;
    MagicTransit.GRETunnels = GRETunnelsAPI.GRETunnels;
    MagicTransit.IPSECTunnels = IPSECTunnelsAPI.IPSECTunnels;
    MagicTransit.Routes = RoutesAPI.Routes;
    MagicTransit.Sites = SitesAPI.Sites;
    MagicTransit.SitesSinglePage = SitesAPI.SitesSinglePage;
    MagicTransit.Connectors = ConnectorsAPI.Connectors;
    MagicTransit.ConnectorListResponsesSinglePage = ConnectorsAPI.ConnectorListResponsesSinglePage;
})(MagicTransit || (MagicTransit = {}));
