// cloudflare@3.5.0/resources/email-routing/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmailRouting = exports.DNS = exports.Addresses = exports.AddressesV4PagePaginationArray = exports.Rules = exports.EmailRoutingRulesV4PagePaginationArray = void 0;
var index_1 = require("./rules/index.js");
Object.defineProperty(exports, "EmailRoutingRulesV4PagePaginationArray", { enumerable: true, get: function () { return index_1.EmailRoutingRulesV4PagePaginationArray; } });
Object.defineProperty(exports, "Rules", { enumerable: true, get: function () { return index_1.Rules; } });
var addresses_1 = require("./addresses.js");
Object.defineProperty(exports, "AddressesV4PagePaginationArray", { enumerable: true, get: function () { return addresses_1.AddressesV4PagePaginationArray; } });
Object.defineProperty(exports, "Addresses", { enumerable: true, get: function () { return addresses_1.Addresses; } });
var dns_1 = require("./dns.js");
Object.defineProperty(exports, "DNS", { enumerable: true, get: function () { return dns_1.DNS; } });
var email_routing_1 = require("./email-routing.js");
Object.defineProperty(exports, "EmailRouting", { enumerable: true, get: function () { return email_routing_1.EmailRouting; } });
