// cloudflare@3.5.0/resources/zero-trust/tunnels/tunnels.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TunnelListResponsesV4PagePaginationArray = exports.Tunnels = void 0;
const resource_1 = require("../../../resource.js");
const TunnelsAPI = __importStar(require("./tunnels.js"));
const ConfigurationsAPI = __importStar(require("./configurations.js"));
const ConnectionsAPI = __importStar(require("./connections.js"));
const ConnectorsAPI = __importStar(require("./connectors.js"));
const ManagementAPI = __importStar(require("./management.js"));
const TokenAPI = __importStar(require("./token.js"));
const pagination_1 = require("../../../pagination.js");
class Tunnels extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.configurations = new ConfigurationsAPI.Configurations(this._client);
        this.connections = new ConnectionsAPI.Connections(this._client);
        this.token = new TokenAPI.Token(this._client);
        this.connectors = new ConnectorsAPI.Connectors(this._client);
        this.management = new ManagementAPI.Management(this._client);
    }
    /**
     * Creates a new Argo Tunnel in an account.
     */
    create(params, options) {
        const { account_id, ...body } = params;
        return this._client.post(`/accounts/${account_id}/tunnels`, { body, ...options })._thenUnwrap((obj) => obj.result);
    }
    /**
     * Lists and filters all types of Tunnels in an account.
     */
    list(params, options) {
        const { account_id, ...query } = params;
        return this._client.getAPIList(`/accounts/${account_id}/tunnels`, TunnelListResponsesV4PagePaginationArray, { query, ...options });
    }
    /**
     * Deletes an Argo Tunnel from an account.
     */
    delete(tunnelId, params, options) {
        const { account_id } = params;
        return this._client.delete(`/accounts/${account_id}/tunnels/${tunnelId}`, options)._thenUnwrap((obj) => obj.result);
    }
    /**
     * Updates an existing Cloudflare Tunnel.
     */
    edit(tunnelId, params, options) {
        const { account_id, ...body } = params;
        return this._client.patch(`/accounts/${account_id}/cfd_tunnel/${tunnelId}`, {
            body,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    /**
     * Fetches a single Argo Tunnel.
     */
    get(tunnelId, params, options) {
        const { account_id } = params;
        return this._client.get(`/accounts/${account_id}/tunnels/${tunnelId}`, options)._thenUnwrap((obj) => obj.result);
    }
}
exports.Tunnels = Tunnels;
class TunnelListResponsesV4PagePaginationArray extends pagination_1.V4PagePaginationArray {
}
exports.TunnelListResponsesV4PagePaginationArray = TunnelListResponsesV4PagePaginationArray;
(function (Tunnels) {
    Tunnels.TunnelListResponsesV4PagePaginationArray = TunnelsAPI.TunnelListResponsesV4PagePaginationArray;
    Tunnels.Configurations = ConfigurationsAPI.Configurations;
    Tunnels.Connections = ConnectionsAPI.Connections;
    Tunnels.Token = TokenAPI.Token;
    Tunnels.Connectors = ConnectorsAPI.Connectors;
    Tunnels.Management = ManagementAPI.Management;
})(Tunnels = exports.Tunnels || (exports.Tunnels = {}));
