// unhead@2.1.15/dist/legacy.mjs
import { a as createUnhead } from './shared/unhead.CaI5ZD4O.mjs';
import { D as DeprecationsPlugin, P as PromisesPlugin, T as TemplateParamsPlugin, A as AliasSortingPlugin } from './shared/unhead.TxTMM7cM.mjs';
export { u as useHead, a as useHeadSafe, b as useSeoMeta, c as useServerHead, d as useServerHeadSafe, e as useServerSeoMeta } from './shared/unhead.Dh1cQCTe.mjs';
export { u as useScript } from './shared/unhead.BnoAbrHA.mjs';
import 'hookable';
import './shared/unhead.B3VbJo7Y.mjs';
import './shared/unhead.yem5I2v_.mjs';
import './shared/unhead.CbpEuj3y.mjs';
import './shared/unhead.D4vSFytZ.mjs';
import './shared/unhead.DQc16pHI.mjs';
import './shared/unhead.BYvz9V1x.mjs';

const activeHead = { value: null };
function getActiveHead() {
  return activeHead?.value;
}
function createServerHead(options = {}) {
  return activeHead.value = createUnhead({
    disableCapoSorting: true,
    ...options,
    // @ts-expect-error untyped
    document: false,
    plugins: [
      ...options.plugins || [],
      DeprecationsPlugin,
      PromisesPlugin,
      TemplateParamsPlugin,
      AliasSortingPlugin
    ]
  });
}
function createHead(options = {}) {
  return activeHead.value = createUnhead({
    disableCapoSorting: true,
    ...options,
    plugins: [
      ...options.plugins || [],
      DeprecationsPlugin,
      PromisesPlugin,
      TemplateParamsPlugin,
      AliasSortingPlugin
    ]
  });
}
const createHeadCore = createUnhead;

export { activeHead, createHead, createHeadCore, createServerHead, createUnhead, getActiveHead };
