// cloudflare@3.5.0/resources/pages/projects/domains.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../resource.mjs";
import * as DomainsAPI from "./domains.mjs";
import { SinglePage } from "../../../pagination.mjs";
export class Domains extends APIResource {
    /**
     * Add a new domain for the Pages project.
     */
    create(projectName, params, options) {
        const { account_id, body } = params;
        return this._client.post(`/accounts/${account_id}/pages/projects/${projectName}/domains`, {
            body: body,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    /**
     * Fetch a list of all domains associated with a Pages project.
     */
    list(projectName, params, options) {
        const { account_id } = params;
        return this._client.getAPIList(`/accounts/${account_id}/pages/projects/${projectName}/domains`, DomainListResponsesSinglePage, options);
    }
    /**
     * Delete a Pages project's domain.
     */
    delete(projectName, domainName, params, options) {
        const { account_id } = params;
        return this._client.delete(`/accounts/${account_id}/pages/projects/${projectName}/domains/${domainName}`, options);
    }
    /**
     * Retry the validation status of a single domain.
     */
    edit(projectName, domainName, params, options) {
        const { account_id, body } = params;
        return this._client.patch(`/accounts/${account_id}/pages/projects/${projectName}/domains/${domainName}`, {
            body: body,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    /**
     * Fetch a single domain.
     */
    get(projectName, domainName, params, options) {
        const { account_id } = params;
        return this._client.get(`/accounts/${account_id}/pages/projects/${projectName}/domains/${domainName}`, options)._thenUnwrap((obj) => obj.result);
    }
}
export class DomainListResponsesSinglePage extends SinglePage {
}
(function (Domains) {
    Domains.DomainListResponsesSinglePage = DomainsAPI.DomainListResponsesSinglePage;
})(Domains || (Domains = {}));
