// @smithy/core@3.25.1/dist-es/submodules/client/index.js
export { constructStack } from "./middleware-stack/MiddlewareStack";
export { getSmithyContext } from "@smithy/core/transport";
export { normalizeProvider } from "@smithy/core/transport";
export { invalidFunction } from "./invalid-dependency/invalidFunction";
export { invalidProvider } from "./invalid-dependency/invalidProvider";
export { createWaiter } from "./util-waiter/createWaiter";
export { waiterServiceDefaults, WaiterState, checkExceptions, } from "./util-waiter/waiter";
export { Client } from "./smithy-client/client";
export { Command } from "./smithy-client/command";
export { SENSITIVE_STRING } from "./smithy-client/constants";
export { createAggregatedClient } from "./smithy-client/create-aggregated-client";
export { throwDefaultError, withBaseException } from "./smithy-client/default-error-handler";
export { loadConfigsForDefaultMode, } from "./smithy-client/defaults-mode";
export { emitWarningIfUnsupportedVersion } from "./smithy-client/emitWarningIfUnsupportedVersion";
export { ServiceException, decorateServiceException, } from "./smithy-client/exceptions";
export { getDefaultExtensionConfiguration, getDefaultClientConfiguration, resolveDefaultRuntimeConfig, } from "./smithy-client/extensions/defaultExtensionConfiguration";
export { AlgorithmId, getChecksumConfiguration, resolveChecksumRuntimeConfig, } from "./smithy-client/extensions/checksum";
export { getRetryConfiguration, resolveRetryRuntimeConfig, } from "./smithy-client/extensions/retry";
export { getArrayIfSingleItem } from "./smithy-client/get-array-if-single-item";
export { getValueFromTextNode } from "./smithy-client/get-value-from-text-node";
export { isSerializableHeaderValue } from "./smithy-client/is-serializable-header-value";
export { NoOpLogger } from "./smithy-client/NoOpLogger";
export { map, convertMap, take, } from "./smithy-client/object-mapping";
export { schemaLogFilter } from "./smithy-client/schemaLogFilter";
export { serializeFloat, serializeDateTime } from "./smithy-client/ser-utils";
export { _json } from "./smithy-client/serde-json";
