// cloudflare@3.5.0/resources/addressing/addressing.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../resource.mjs";
import * as ServicesAPI from "./services.mjs";
import * as AddressMapsAPI from "./address-maps/address-maps.mjs";
import * as LOADocumentsAPI from "./loa-documents/loa-documents.mjs";
import * as PrefixesAPI from "./prefixes/prefixes.mjs";
import * as RegionalHostnamesAPI from "./regional-hostnames/regional-hostnames.mjs";
export class Addressing extends APIResource {
    constructor() {
        super(...arguments);
        this.regionalHostnames = new RegionalHostnamesAPI.RegionalHostnames(this._client);
        this.services = new ServicesAPI.Services(this._client);
        this.addressMaps = new AddressMapsAPI.AddressMaps(this._client);
        this.loaDocuments = new LOADocumentsAPI.LOADocuments(this._client);
        this.prefixes = new PrefixesAPI.Prefixes(this._client);
    }
}
(function (Addressing) {
    Addressing.RegionalHostnames = RegionalHostnamesAPI.RegionalHostnames;
    Addressing.RegionalHostnameListResponsesSinglePage = RegionalHostnamesAPI.RegionalHostnameListResponsesSinglePage;
    Addressing.Services = ServicesAPI.Services;
    Addressing.ServiceListResponsesSinglePage = ServicesAPI.ServiceListResponsesSinglePage;
    Addressing.AddressMaps = AddressMapsAPI.AddressMaps;
    Addressing.AddressMapsSinglePage = AddressMapsAPI.AddressMapsSinglePage;
    Addressing.LOADocuments = LOADocumentsAPI.LOADocuments;
    Addressing.Prefixes = PrefixesAPI.Prefixes;
    Addressing.PrefixesSinglePage = PrefixesAPI.PrefixesSinglePage;
})(Addressing || (Addressing = {}));
