// cloudflare@3.5.0/resources/secondary-dns/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TSIGs = exports.TSIGsSinglePage = exports.SecondaryDNS = exports.Peers = exports.PeersSinglePage = exports.IncomingResource = exports.ForceAXFRResource = exports.OutgoingResource = exports.ACLs = exports.ACLsSinglePage = void 0;
var acls_1 = require("./acls.js");
Object.defineProperty(exports, "ACLsSinglePage", { enumerable: true, get: function () { return acls_1.ACLsSinglePage; } });
Object.defineProperty(exports, "ACLs", { enumerable: true, get: function () { return acls_1.ACLs; } });
var index_1 = require("./outgoing/index.js");
Object.defineProperty(exports, "OutgoingResource", { enumerable: true, get: function () { return index_1.OutgoingResource; } });
var force_axfr_1 = require("./force-axfr.js");
Object.defineProperty(exports, "ForceAXFRResource", { enumerable: true, get: function () { return force_axfr_1.ForceAXFRResource; } });
var incoming_1 = require("./incoming.js");
Object.defineProperty(exports, "IncomingResource", { enumerable: true, get: function () { return incoming_1.IncomingResource; } });
var peers_1 = require("./peers.js");
Object.defineProperty(exports, "PeersSinglePage", { enumerable: true, get: function () { return peers_1.PeersSinglePage; } });
Object.defineProperty(exports, "Peers", { enumerable: true, get: function () { return peers_1.Peers; } });
var secondary_dns_1 = require("./secondary-dns.js");
Object.defineProperty(exports, "SecondaryDNS", { enumerable: true, get: function () { return secondary_dns_1.SecondaryDNS; } });
var tsigs_1 = require("./tsigs.js");
Object.defineProperty(exports, "TSIGsSinglePage", { enumerable: true, get: function () { return tsigs_1.TSIGsSinglePage; } });
Object.defineProperty(exports, "TSIGs", { enumerable: true, get: function () { return tsigs_1.TSIGs; } });
