// cloudflare@3.5.0/resources/secondary-dns/peers.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../resource.mjs";
import * as PeersAPI from "./peers.mjs";
import { SinglePage } from "../../pagination.mjs";
export class Peers extends APIResource {
    /**
     * Create Peer.
     */
    create(params, options) {
        const { account_id, body } = params;
        return this._client.post(`/accounts/${account_id}/secondary_dns/peers`, {
            body: body,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    /**
     * Modify Peer.
     */
    update(peerId, params, options) {
        const { account_id, ...body } = params;
        return this._client.put(`/accounts/${account_id}/secondary_dns/peers/${peerId}`, {
            body,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    /**
     * List Peers.
     */
    list(params, options) {
        const { account_id } = params;
        return this._client.getAPIList(`/accounts/${account_id}/secondary_dns/peers`, PeersSinglePage, options);
    }
    /**
     * Delete Peer.
     */
    delete(peerId, params, options) {
        const { account_id } = params;
        return this._client.delete(`/accounts/${account_id}/secondary_dns/peers/${peerId}`, options)._thenUnwrap((obj) => obj.result);
    }
    /**
     * Get Peer.
     */
    get(peerId, params, options) {
        const { account_id } = params;
        return this._client.get(`/accounts/${account_id}/secondary_dns/peers/${peerId}`, options)._thenUnwrap((obj) => obj.result);
    }
}
export class PeersSinglePage extends SinglePage {
}
(function (Peers) {
    Peers.PeersSinglePage = PeersAPI.PeersSinglePage;
})(Peers || (Peers = {}));
