// cloudflare@3.5.0/resources/secondary-dns/secondary-dns.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../resource.mjs";
import * as ACLsAPI from "./acls.mjs";
import * as ForceAXFRAPI from "./force-axfr.mjs";
import * as IncomingAPI from "./incoming.mjs";
import * as PeersAPI from "./peers.mjs";
import * as TSIGsAPI from "./tsigs.mjs";
import * as OutgoingAPI from "./outgoing/outgoing.mjs";
export class SecondaryDNS extends APIResource {
    constructor() {
        super(...arguments);
        this.forceAXFR = new ForceAXFRAPI.ForceAXFRResource(this._client);
        this.incoming = new IncomingAPI.IncomingResource(this._client);
        this.outgoing = new OutgoingAPI.OutgoingResource(this._client);
        this.acls = new ACLsAPI.ACLs(this._client);
        this.peers = new PeersAPI.Peers(this._client);
        this.tsigs = new TSIGsAPI.TSIGs(this._client);
    }
}
(function (SecondaryDNS) {
    SecondaryDNS.ForceAXFRResource = ForceAXFRAPI.ForceAXFRResource;
    SecondaryDNS.IncomingResource = IncomingAPI.IncomingResource;
    SecondaryDNS.OutgoingResource = OutgoingAPI.OutgoingResource;
    SecondaryDNS.ACLs = ACLsAPI.ACLs;
    SecondaryDNS.ACLsSinglePage = ACLsAPI.ACLsSinglePage;
    SecondaryDNS.Peers = PeersAPI.Peers;
    SecondaryDNS.PeersSinglePage = PeersAPI.PeersSinglePage;
    SecondaryDNS.TSIGs = TSIGsAPI.TSIGs;
    SecondaryDNS.TSIGsSinglePage = TSIGsAPI.TSIGsSinglePage;
})(SecondaryDNS || (SecondaryDNS = {}));
