// cloudflare@3.5.0/resources/zero-trust/access/applications/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPolicyChecks = exports.CAs = exports.CAsSinglePage = exports.Policies = exports.PolicyListResponsesSinglePage = exports.Applications = exports.ApplicationListResponsesSinglePage = void 0;
var applications_1 = require("./applications.js");
Object.defineProperty(exports, "ApplicationListResponsesSinglePage", { enumerable: true, get: function () { return applications_1.ApplicationListResponsesSinglePage; } });
Object.defineProperty(exports, "Applications", { enumerable: true, get: function () { return applications_1.Applications; } });
var policies_1 = require("./policies.js");
Object.defineProperty(exports, "PolicyListResponsesSinglePage", { enumerable: true, get: function () { return policies_1.PolicyListResponsesSinglePage; } });
Object.defineProperty(exports, "Policies", { enumerable: true, get: function () { return policies_1.Policies; } });
var cas_1 = require("./cas.js");
Object.defineProperty(exports, "CAsSinglePage", { enumerable: true, get: function () { return cas_1.CAsSinglePage; } });
Object.defineProperty(exports, "CAs", { enumerable: true, get: function () { return cas_1.CAs; } });
var user_policy_checks_1 = require("./user-policy-checks.js");
Object.defineProperty(exports, "UserPolicyChecks", { enumerable: true, get: function () { return user_policy_checks_1.UserPolicyChecks; } });
