// @smithy/core@3.25.1/dist-es/submodules/event-streams/index.js
export { EventStreamCodec } from "./eventstream-codec/EventStreamCodec";
export { HeaderMarshaller } from "./eventstream-codec/HeaderMarshaller";
export { Int64 } from "./eventstream-codec/Int64";
export { MessageDecoderStream } from "./eventstream-codec/MessageDecoderStream";
export { MessageEncoderStream } from "./eventstream-codec/MessageEncoderStream";
export { SmithyMessageDecoderStream } from "./eventstream-codec/SmithyMessageDecoderStream";
export { SmithyMessageEncoderStream } from "./eventstream-codec/SmithyMessageEncoderStream";
export { EventStreamMarshaller, eventStreamSerdeProvider } from "./eventstream-serde/EventStreamMarshaller";
export { readableStreamToIterable, iterableToReadableStream } from "./eventstream-serde/utils";
export { EventStreamMarshaller as UniversalEventStreamMarshaller, eventStreamSerdeProvider as universalEventStreamSerdeProvider, } from "./eventstream-serde-universal/EventStreamMarshaller";
export { getChunkedStream } from "./eventstream-serde-universal/getChunkedStream";
export { getUnmarshalledStream, getMessageUnmarshaller } from "./eventstream-serde-universal/getUnmarshalledStream";
export { resolveEventStreamSerdeConfig } from "./eventstream-serde-config-resolver/EventStreamSerdeConfig";
export { EventStreamSerde } from "./EventStreamSerde";
