// cloudflare@3.5.0/resources/zero-trust/dex/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TracerouteTestResults = exports.TracerouteTests = exports.FleetStatus = exports.HTTPTests = exports.DEX = exports.Colos = exports.ColoListResponsesSinglePage = exports.TestListResponsesV4PagePagination = exports.Tests = void 0;
var index_1 = require("./tests/index.js");
Object.defineProperty(exports, "Tests", { enumerable: true, get: function () { return index_1.Tests; } });
Object.defineProperty(exports, "TestListResponsesV4PagePagination", { enumerable: true, get: function () { return index_1.TestListResponsesV4PagePagination; } });
var colos_1 = require("./colos.js");
Object.defineProperty(exports, "ColoListResponsesSinglePage", { enumerable: true, get: function () { return colos_1.ColoListResponsesSinglePage; } });
Object.defineProperty(exports, "Colos", { enumerable: true, get: function () { return colos_1.Colos; } });
var dex_1 = require("./dex.js");
Object.defineProperty(exports, "DEX", { enumerable: true, get: function () { return dex_1.DEX; } });
var index_2 = require("./http-tests/index.js");
Object.defineProperty(exports, "HTTPTests", { enumerable: true, get: function () { return index_2.HTTPTests; } });
var index_3 = require("./fleet-status/index.js");
Object.defineProperty(exports, "FleetStatus", { enumerable: true, get: function () { return index_3.FleetStatus; } });
var traceroute_tests_1 = require("./traceroute-tests.js");
Object.defineProperty(exports, "TracerouteTests", { enumerable: true, get: function () { return traceroute_tests_1.TracerouteTests; } });
var index_4 = require("./traceroute-test-results/index.js");
Object.defineProperty(exports, "TracerouteTestResults", { enumerable: true, get: function () { return index_4.TracerouteTestResults; } });
