// cloudflare@3.5.0/resources/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OriginPostQuantumEncryption = exports.OriginCACertificates = exports.Memberships = exports.ManagedHeaders = exports.MagicTransit = exports.MagicNetworkMonitoring = exports.MTLSCertificates = exports.Logs = exports.Logpush = exports.LoadBalancers = exports.KeylessCertificates = exports.KV = exports.Intel = exports.Images = exports.IPs = exports.IAM = exports.HyperdriveResource = exports.Hostnames = exports.Healthchecks = exports.Firewall = exports.Filters = exports.EventNotifications = exports.EmailRouting = exports.DurableObjects = exports.Diagnostics = exports.DNSSECResource = exports.DNS = exports.DCVDelegation = exports.D1Resource = exports.CustomNameservers = exports.CustomHostnames = exports.CustomCertificates = exports.CloudforceOne = exports.CloudConnector = exports.ClientCertificates = exports.Challenges = exports.CertificateAuthorities = exports.Calls = exports.Cache = exports.BrandProtection = exports.BotManagement = exports.Billing = exports.AuditLogs = exports.Argo = exports.Alerting = exports.Addressing = exports.Accounts = exports.APIGateway = exports.AIGateway = exports.ACM = void 0;
exports.Zones = exports.ZeroTrust = exports.WorkersForPlatforms = exports.Workers = exports.Web3 = exports.WaitingRooms = exports.WARPConnector = exports.Vectorize = exports.User = exports.URLScanner = exports.URLNormalization = exports.Subscriptions = exports.Stream = exports.Storage = exports.Speed = exports.Spectrum = exports.Snippets = exports.SecondaryDNS = exports.SSL = exports.Rulesets = exports.Rules = exports.RequestTracers = exports.Registrar = exports.RatePlans = exports.RateLimits = exports.Radar = exports.RUM = exports.R2 = exports.Queues = exports.Plans = exports.Pages = exports.Pagerules = exports.PageShield = exports.PCAPs = exports.OriginTLSClientAuth = void 0;
__exportStar(require("./shared.js"), exports);
var acm_1 = require("./acm/acm.js");
Object.defineProperty(exports, "ACM", { enumerable: true, get: function () { return acm_1.ACM; } });
var ai_gateway_1 = require("./ai-gateway/ai-gateway.js");
Object.defineProperty(exports, "AIGateway", { enumerable: true, get: function () { return ai_gateway_1.AIGateway; } });
var api_gateway_1 = require("./api-gateway/api-gateway.js");
Object.defineProperty(exports, "APIGateway", { enumerable: true, get: function () { return api_gateway_1.APIGateway; } });
var accounts_1 = require("./accounts/accounts.js");
Object.defineProperty(exports, "Accounts", { enumerable: true, get: function () { return accounts_1.Accounts; } });
var addressing_1 = require("./addressing/addressing.js");
Object.defineProperty(exports, "Addressing", { enumerable: true, get: function () { return addressing_1.Addressing; } });
var alerting_1 = require("./alerting/alerting.js");
Object.defineProperty(exports, "Alerting", { enumerable: true, get: function () { return alerting_1.Alerting; } });
var argo_1 = require("./argo/argo.js");
Object.defineProperty(exports, "Argo", { enumerable: true, get: function () { return argo_1.Argo; } });
var audit_logs_1 = require("./audit-logs.js");
Object.defineProperty(exports, "AuditLogs", { enumerable: true, get: function () { return audit_logs_1.AuditLogs; } });
var billing_1 = require("./billing/billing.js");
Object.defineProperty(exports, "Billing", { enumerable: true, get: function () { return billing_1.Billing; } });
var bot_management_1 = require("./bot-management.js");
Object.defineProperty(exports, "BotManagement", { enumerable: true, get: function () { return bot_management_1.BotManagement; } });
var brand_protection_1 = require("./brand-protection.js");
Object.defineProperty(exports, "BrandProtection", { enumerable: true, get: function () { return brand_protection_1.BrandProtection; } });
var cache_1 = require("./cache/cache.js");
Object.defineProperty(exports, "Cache", { enumerable: true, get: function () { return cache_1.Cache; } });
var calls_1 = require("./calls/calls.js");
Object.defineProperty(exports, "Calls", { enumerable: true, get: function () { return calls_1.Calls; } });
var certificate_authorities_1 = require("./certificate-authorities/certificate-authorities.js");
Object.defineProperty(exports, "CertificateAuthorities", { enumerable: true, get: function () { return certificate_authorities_1.CertificateAuthorities; } });
var challenges_1 = require("./challenges/challenges.js");
Object.defineProperty(exports, "Challenges", { enumerable: true, get: function () { return challenges_1.Challenges; } });
var client_certificates_1 = require("./client-certificates.js");
Object.defineProperty(exports, "ClientCertificates", { enumerable: true, get: function () { return client_certificates_1.ClientCertificates; } });
var cloud_connector_1 = require("./cloud-connector/cloud-connector.js");
Object.defineProperty(exports, "CloudConnector", { enumerable: true, get: function () { return cloud_connector_1.CloudConnector; } });
var cloudforce_one_1 = require("./cloudforce-one/cloudforce-one.js");
Object.defineProperty(exports, "CloudforceOne", { enumerable: true, get: function () { return cloudforce_one_1.CloudforceOne; } });
var custom_certificates_1 = require("./custom-certificates/custom-certificates.js");
Object.defineProperty(exports, "CustomCertificates", { enumerable: true, get: function () { return custom_certificates_1.CustomCertificates; } });
var custom_hostnames_1 = require("./custom-hostnames/custom-hostnames.js");
Object.defineProperty(exports, "CustomHostnames", { enumerable: true, get: function () { return custom_hostnames_1.CustomHostnames; } });
var custom_nameservers_1 = require("./custom-nameservers.js");
Object.defineProperty(exports, "CustomNameservers", { enumerable: true, get: function () { return custom_nameservers_1.CustomNameservers; } });
var d1_1 = require("./d1/d1.js");
Object.defineProperty(exports, "D1Resource", { enumerable: true, get: function () { return d1_1.D1Resource; } });
var dcv_delegation_1 = require("./dcv-delegation.js");
Object.defineProperty(exports, "DCVDelegation", { enumerable: true, get: function () { return dcv_delegation_1.DCVDelegation; } });
var dns_1 = require("./dns/dns.js");
Object.defineProperty(exports, "DNS", { enumerable: true, get: function () { return dns_1.DNS; } });
var dnssec_1 = require("./dnssec.js");
Object.defineProperty(exports, "DNSSECResource", { enumerable: true, get: function () { return dnssec_1.DNSSECResource; } });
var diagnostics_1 = require("./diagnostics/diagnostics.js");
Object.defineProperty(exports, "Diagnostics", { enumerable: true, get: function () { return diagnostics_1.Diagnostics; } });
var durable_objects_1 = require("./durable-objects/durable-objects.js");
Object.defineProperty(exports, "DurableObjects", { enumerable: true, get: function () { return durable_objects_1.DurableObjects; } });
var email_routing_1 = require("./email-routing/email-routing.js");
Object.defineProperty(exports, "EmailRouting", { enumerable: true, get: function () { return email_routing_1.EmailRouting; } });
var event_notifications_1 = require("./event-notifications/event-notifications.js");
Object.defineProperty(exports, "EventNotifications", { enumerable: true, get: function () { return event_notifications_1.EventNotifications; } });
var filters_1 = require("./filters.js");
Object.defineProperty(exports, "Filters", { enumerable: true, get: function () { return filters_1.Filters; } });
var firewall_1 = require("./firewall/firewall.js");
Object.defineProperty(exports, "Firewall", { enumerable: true, get: function () { return firewall_1.Firewall; } });
var healthchecks_1 = require("./healthchecks/healthchecks.js");
Object.defineProperty(exports, "Healthchecks", { enumerable: true, get: function () { return healthchecks_1.Healthchecks; } });
var hostnames_1 = require("./hostnames/hostnames.js");
Object.defineProperty(exports, "Hostnames", { enumerable: true, get: function () { return hostnames_1.Hostnames; } });
var hyperdrive_1 = require("./hyperdrive/hyperdrive.js");
Object.defineProperty(exports, "HyperdriveResource", { enumerable: true, get: function () { return hyperdrive_1.HyperdriveResource; } });
var iam_1 = require("./iam/iam.js");
Object.defineProperty(exports, "IAM", { enumerable: true, get: function () { return iam_1.IAM; } });
var ips_1 = require("./ips.js");
Object.defineProperty(exports, "IPs", { enumerable: true, get: function () { return ips_1.IPs; } });
var images_1 = require("./images/images.js");
Object.defineProperty(exports, "Images", { enumerable: true, get: function () { return images_1.Images; } });
var intel_1 = require("./intel/intel.js");
Object.defineProperty(exports, "Intel", { enumerable: true, get: function () { return intel_1.Intel; } });
var kv_1 = require("./kv/kv.js");
Object.defineProperty(exports, "KV", { enumerable: true, get: function () { return kv_1.KV; } });
var keyless_certificates_1 = require("./keyless-certificates.js");
Object.defineProperty(exports, "KeylessCertificates", { enumerable: true, get: function () { return keyless_certificates_1.KeylessCertificates; } });
var load_balancers_1 = require("./load-balancers/load-balancers.js");
Object.defineProperty(exports, "LoadBalancers", { enumerable: true, get: function () { return load_balancers_1.LoadBalancers; } });
var logpush_1 = require("./logpush/logpush.js");
Object.defineProperty(exports, "Logpush", { enumerable: true, get: function () { return logpush_1.Logpush; } });
var logs_1 = require("./logs/logs.js");
Object.defineProperty(exports, "Logs", { enumerable: true, get: function () { return logs_1.Logs; } });
var mtls_certificates_1 = require("./mtls-certificates/mtls-certificates.js");
Object.defineProperty(exports, "MTLSCertificates", { enumerable: true, get: function () { return mtls_certificates_1.MTLSCertificates; } });
var magic_network_monitoring_1 = require("./magic-network-monitoring/magic-network-monitoring.js");
Object.defineProperty(exports, "MagicNetworkMonitoring", { enumerable: true, get: function () { return magic_network_monitoring_1.MagicNetworkMonitoring; } });
var magic_transit_1 = require("./magic-transit/magic-transit.js");
Object.defineProperty(exports, "MagicTransit", { enumerable: true, get: function () { return magic_transit_1.MagicTransit; } });
var managed_headers_1 = require("./managed-headers.js");
Object.defineProperty(exports, "ManagedHeaders", { enumerable: true, get: function () { return managed_headers_1.ManagedHeaders; } });
var memberships_1 = require("./memberships.js");
Object.defineProperty(exports, "Memberships", { enumerable: true, get: function () { return memberships_1.Memberships; } });
var origin_ca_certificates_1 = require("./origin-ca-certificates.js");
Object.defineProperty(exports, "OriginCACertificates", { enumerable: true, get: function () { return origin_ca_certificates_1.OriginCACertificates; } });
var origin_post_quantum_encryption_1 = require("./origin-post-quantum-encryption.js");
Object.defineProperty(exports, "OriginPostQuantumEncryption", { enumerable: true, get: function () { return origin_post_quantum_encryption_1.OriginPostQuantumEncryption; } });
var origin_tls_client_auth_1 = require("./origin-tls-client-auth/origin-tls-client-auth.js");
Object.defineProperty(exports, "OriginTLSClientAuth", { enumerable: true, get: function () { return origin_tls_client_auth_1.OriginTLSClientAuth; } });
var pcaps_1 = require("./pcaps/pcaps.js");
Object.defineProperty(exports, "PCAPs", { enumerable: true, get: function () { return pcaps_1.PCAPs; } });
var page_shield_1 = require("./page-shield/page-shield.js");
Object.defineProperty(exports, "PageShield", { enumerable: true, get: function () { return page_shield_1.PageShield; } });
var pagerules_1 = require("./pagerules/pagerules.js");
Object.defineProperty(exports, "Pagerules", { enumerable: true, get: function () { return pagerules_1.Pagerules; } });
var pages_1 = require("./pages/pages.js");
Object.defineProperty(exports, "Pages", { enumerable: true, get: function () { return pages_1.Pages; } });
var plans_1 = require("./plans.js");
Object.defineProperty(exports, "Plans", { enumerable: true, get: function () { return plans_1.Plans; } });
var queues_1 = require("./queues/queues.js");
Object.defineProperty(exports, "Queues", { enumerable: true, get: function () { return queues_1.Queues; } });
var r2_1 = require("./r2/r2.js");
Object.defineProperty(exports, "R2", { enumerable: true, get: function () { return r2_1.R2; } });
var rum_1 = require("./rum/rum.js");
Object.defineProperty(exports, "RUM", { enumerable: true, get: function () { return rum_1.RUM; } });
var radar_1 = require("./radar/radar.js");
Object.defineProperty(exports, "Radar", { enumerable: true, get: function () { return radar_1.Radar; } });
var rate_limits_1 = require("./rate-limits.js");
Object.defineProperty(exports, "RateLimits", { enumerable: true, get: function () { return rate_limits_1.RateLimits; } });
var rate_plans_1 = require("./rate-plans.js");
Object.defineProperty(exports, "RatePlans", { enumerable: true, get: function () { return rate_plans_1.RatePlans; } });
var registrar_1 = require("./registrar/registrar.js");
Object.defineProperty(exports, "Registrar", { enumerable: true, get: function () { return registrar_1.Registrar; } });
var request_tracers_1 = require("./request-tracers/request-tracers.js");
Object.defineProperty(exports, "RequestTracers", { enumerable: true, get: function () { return request_tracers_1.RequestTracers; } });
var rules_1 = require("./rules/rules.js");
Object.defineProperty(exports, "Rules", { enumerable: true, get: function () { return rules_1.Rules; } });
var rulesets_1 = require("./rulesets/rulesets.js");
Object.defineProperty(exports, "Rulesets", { enumerable: true, get: function () { return rulesets_1.Rulesets; } });
var ssl_1 = require("./ssl/ssl.js");
Object.defineProperty(exports, "SSL", { enumerable: true, get: function () { return ssl_1.SSL; } });
var secondary_dns_1 = require("./secondary-dns/secondary-dns.js");
Object.defineProperty(exports, "SecondaryDNS", { enumerable: true, get: function () { return secondary_dns_1.SecondaryDNS; } });
var snippets_1 = require("./snippets/snippets.js");
Object.defineProperty(exports, "Snippets", { enumerable: true, get: function () { return snippets_1.Snippets; } });
var spectrum_1 = require("./spectrum/spectrum.js");
Object.defineProperty(exports, "Spectrum", { enumerable: true, get: function () { return spectrum_1.Spectrum; } });
var speed_1 = require("./speed/speed.js");
Object.defineProperty(exports, "Speed", { enumerable: true, get: function () { return speed_1.Speed; } });
var storage_1 = require("./storage/storage.js");
Object.defineProperty(exports, "Storage", { enumerable: true, get: function () { return storage_1.Storage; } });
var stream_1 = require("./stream/stream.js");
Object.defineProperty(exports, "Stream", { enumerable: true, get: function () { return stream_1.Stream; } });
var subscriptions_1 = require("./subscriptions.js");
Object.defineProperty(exports, "Subscriptions", { enumerable: true, get: function () { return subscriptions_1.Subscriptions; } });
var url_normalization_1 = require("./url-normalization.js");
Object.defineProperty(exports, "URLNormalization", { enumerable: true, get: function () { return url_normalization_1.URLNormalization; } });
var url_scanner_1 = require("./url-scanner/url-scanner.js");
Object.defineProperty(exports, "URLScanner", { enumerable: true, get: function () { return url_scanner_1.URLScanner; } });
var user_1 = require("./user/user.js");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return user_1.User; } });
var vectorize_1 = require("./vectorize/vectorize.js");
Object.defineProperty(exports, "Vectorize", { enumerable: true, get: function () { return vectorize_1.Vectorize; } });
var warp_connector_1 = require("./warp-connector.js");
Object.defineProperty(exports, "WARPConnector", { enumerable: true, get: function () { return warp_connector_1.WARPConnector; } });
var waiting_rooms_1 = require("./waiting-rooms/waiting-rooms.js");
Object.defineProperty(exports, "WaitingRooms", { enumerable: true, get: function () { return waiting_rooms_1.WaitingRooms; } });
var web3_1 = require("./web3/web3.js");
Object.defineProperty(exports, "Web3", { enumerable: true, get: function () { return web3_1.Web3; } });
var workers_1 = require("./workers/workers.js");
Object.defineProperty(exports, "Workers", { enumerable: true, get: function () { return workers_1.Workers; } });
var workers_for_platforms_1 = require("./workers-for-platforms/workers-for-platforms.js");
Object.defineProperty(exports, "WorkersForPlatforms", { enumerable: true, get: function () { return workers_for_platforms_1.WorkersForPlatforms; } });
var zero_trust_1 = require("./zero-trust/zero-trust.js");
Object.defineProperty(exports, "ZeroTrust", { enumerable: true, get: function () { return zero_trust_1.ZeroTrust; } });
var zones_1 = require("./zones/zones.js");
Object.defineProperty(exports, "Zones", { enumerable: true, get: function () { return zones_1.Zones; } });
