// cloudflare@3.5.0/resources/calls/turn/keys.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyListResponsesSinglePage = exports.Keys = void 0;
const resource_1 = require("../../../resource.js");
const KeysAPI = __importStar(require("./keys.js"));
const pagination_1 = require("../../../pagination.js");
class Keys extends resource_1.APIResource {
    /**
     * Creates a new Cloudflare Calls TURN key.
     */
    create(params, options) {
        const { account_id, ...body } = params;
        return this._client.post(`/accounts/${account_id}/calls/turn_keys`, { body, ...options });
    }
    /**
     * Edit details for a single TURN key.
     */
    update(keyId, params, options) {
        const { account_id, ...body } = params;
        return this._client.put(`/accounts/${account_id}/calls/turn_keys/${keyId}`, {
            body,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    /**
     * Lists all TURN keys in the Cloudflare account
     */
    list(params, options) {
        const { account_id } = params;
        return this._client.getAPIList(`/accounts/${account_id}/calls/turn_keys`, KeyListResponsesSinglePage, options);
    }
    /**
     * Deletes a TURN key from Cloudflare Calls
     */
    delete(keyId, params, options) {
        const { account_id } = params;
        return this._client.delete(`/accounts/${account_id}/calls/turn_keys/${keyId}`, options)._thenUnwrap((obj) => obj.result);
    }
    /**
     * Fetches details for a single TURN key.
     */
    get(keyId, params, options) {
        const { account_id } = params;
        return this._client.get(`/accounts/${account_id}/calls/turn_keys/${keyId}`, options)._thenUnwrap((obj) => obj.result);
    }
}
exports.Keys = Keys;
class KeyListResponsesSinglePage extends pagination_1.SinglePage {
}
exports.KeyListResponsesSinglePage = KeyListResponsesSinglePage;
(function (Keys) {
    Keys.KeyListResponsesSinglePage = KeysAPI.KeyListResponsesSinglePage;
})(Keys = exports.Keys || (exports.Keys = {}));
