// cloudflare@3.5.0/resources/calls/turn/keys.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../resource.mjs";
import * as KeysAPI from "./keys.mjs";
import { SinglePage } from "../../../pagination.mjs";
export class Keys extends APIResource {
    /**
     * Creates a new Cloudflare Calls TURN key.
     */
    create(params, options) {
        const { account_id, ...body } = params;
        return this._client.post(`/accounts/${account_id}/calls/turn_keys`, { body, ...options });
    }
    /**
     * Edit details for a single TURN key.
     */
    update(keyId, params, options) {
        const { account_id, ...body } = params;
        return this._client.put(`/accounts/${account_id}/calls/turn_keys/${keyId}`, {
            body,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    /**
     * Lists all TURN keys in the Cloudflare account
     */
    list(params, options) {
        const { account_id } = params;
        return this._client.getAPIList(`/accounts/${account_id}/calls/turn_keys`, KeyListResponsesSinglePage, options);
    }
    /**
     * Deletes a TURN key from Cloudflare Calls
     */
    delete(keyId, params, options) {
        const { account_id } = params;
        return this._client.delete(`/accounts/${account_id}/calls/turn_keys/${keyId}`, options)._thenUnwrap((obj) => obj.result);
    }
    /**
     * Fetches details for a single TURN key.
     */
    get(keyId, params, options) {
        const { account_id } = params;
        return this._client.get(`/accounts/${account_id}/calls/turn_keys/${keyId}`, options)._thenUnwrap((obj) => obj.result);
    }
}
export class KeyListResponsesSinglePage extends SinglePage {
}
(function (Keys) {
    Keys.KeyListResponsesSinglePage = KeysAPI.KeyListResponsesSinglePage;
})(Keys || (Keys = {}));
