// cloudflare@3.5.0/resources/zero-trust/devices/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Unrevoke = exports.Policies = exports.SettingsPoliciesSinglePage = exports.Revoke = exports.OverrideCodes = exports.Settings = exports.Networks = exports.DeviceNetworksSinglePage = exports.Devices = exports.DevicesSinglePage = exports.DEXTests = exports.SchemaHTTPSSinglePage = exports.Posture = exports.DevicePostureRulesSinglePage = void 0;
var index_1 = require("./posture/index.js");
Object.defineProperty(exports, "DevicePostureRulesSinglePage", { enumerable: true, get: function () { return index_1.DevicePostureRulesSinglePage; } });
Object.defineProperty(exports, "Posture", { enumerable: true, get: function () { return index_1.Posture; } });
var dex_tests_1 = require("./dex-tests.js");
Object.defineProperty(exports, "SchemaHTTPSSinglePage", { enumerable: true, get: function () { return dex_tests_1.SchemaHTTPSSinglePage; } });
Object.defineProperty(exports, "DEXTests", { enumerable: true, get: function () { return dex_tests_1.DEXTests; } });
var devices_1 = require("./devices.js");
Object.defineProperty(exports, "DevicesSinglePage", { enumerable: true, get: function () { return devices_1.DevicesSinglePage; } });
Object.defineProperty(exports, "Devices", { enumerable: true, get: function () { return devices_1.Devices; } });
var networks_1 = require("./networks.js");
Object.defineProperty(exports, "DeviceNetworksSinglePage", { enumerable: true, get: function () { return networks_1.DeviceNetworksSinglePage; } });
Object.defineProperty(exports, "Networks", { enumerable: true, get: function () { return networks_1.Networks; } });
var settings_1 = require("./settings.js");
Object.defineProperty(exports, "Settings", { enumerable: true, get: function () { return settings_1.Settings; } });
var override_codes_1 = require("./override-codes.js");
Object.defineProperty(exports, "OverrideCodes", { enumerable: true, get: function () { return override_codes_1.OverrideCodes; } });
var revoke_1 = require("./revoke.js");
Object.defineProperty(exports, "Revoke", { enumerable: true, get: function () { return revoke_1.Revoke; } });
var index_2 = require("./policies/index.js");
Object.defineProperty(exports, "SettingsPoliciesSinglePage", { enumerable: true, get: function () { return index_2.SettingsPoliciesSinglePage; } });
Object.defineProperty(exports, "Policies", { enumerable: true, get: function () { return index_2.Policies; } });
var unrevoke_1 = require("./unrevoke.js");
Object.defineProperty(exports, "Unrevoke", { enumerable: true, get: function () { return unrevoke_1.Unrevoke; } });
