// cloudflare@3.5.0/resources/zero-trust/dlp/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PayloadLogs = exports.Patterns = exports.Datasets = exports.DatasetsSinglePage = exports.DLP = exports.Profiles = exports.ProfilesSinglePage = void 0;
var index_1 = require("./profiles/index.js");
Object.defineProperty(exports, "ProfilesSinglePage", { enumerable: true, get: function () { return index_1.ProfilesSinglePage; } });
Object.defineProperty(exports, "Profiles", { enumerable: true, get: function () { return index_1.Profiles; } });
var dlp_1 = require("./dlp.js");
Object.defineProperty(exports, "DLP", { enumerable: true, get: function () { return dlp_1.DLP; } });
var index_2 = require("./datasets/index.js");
Object.defineProperty(exports, "DatasetsSinglePage", { enumerable: true, get: function () { return index_2.DatasetsSinglePage; } });
Object.defineProperty(exports, "Datasets", { enumerable: true, get: function () { return index_2.Datasets; } });
var patterns_1 = require("./patterns.js");
Object.defineProperty(exports, "Patterns", { enumerable: true, get: function () { return patterns_1.Patterns; } });
var payload_logs_1 = require("./payload-logs.js");
Object.defineProperty(exports, "PayloadLogs", { enumerable: true, get: function () { return payload_logs_1.PayloadLogs; } });
