// cloudflare@3.5.0/resources/spectrum/apps.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppListResponsesV4PagePaginationArray = exports.Apps = void 0;
const resource_1 = require("../../resource.js");
const core_1 = require("../../core.js");
const AppsAPI = __importStar(require("./apps.js"));
const pagination_1 = require("../../pagination.js");
class Apps extends resource_1.APIResource {
    /**
     * Creates a new Spectrum application from a configuration using a name for the
     * origin.
     */
    create(zone, body, options) {
        return this._client.post(`/zones/${zone}/spectrum/apps`, { body, ...options })._thenUnwrap((obj) => obj.result);
    }
    /**
     * Updates a previously existing application's configuration that uses a name for
     * the origin.
     */
    update(zone, appId, body, options) {
        return this._client.put(`/zones/${zone}/spectrum/apps/${appId}`, { body, ...options })._thenUnwrap((obj) => obj.result);
    }
    list(zone, query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list(zone, {}, query);
        }
        return this._client.getAPIList(`/zones/${zone}/spectrum/apps`, AppListResponsesV4PagePaginationArray, {
            query,
            ...options,
        });
    }
    /**
     * Deletes a previously existing application.
     */
    delete(zone, appId, options) {
        return this._client.delete(`/zones/${zone}/spectrum/apps/${appId}`, options)._thenUnwrap((obj) => obj.result);
    }
    /**
     * Gets the application configuration of a specific application inside a zone.
     */
    get(zone, appId, options) {
        return this._client.get(`/zones/${zone}/spectrum/apps/${appId}`, options)._thenUnwrap((obj) => obj.result);
    }
}
exports.Apps = Apps;
class AppListResponsesV4PagePaginationArray extends pagination_1.V4PagePaginationArray {
}
exports.AppListResponsesV4PagePaginationArray = AppListResponsesV4PagePaginationArray;
(function (Apps) {
    Apps.AppListResponsesV4PagePaginationArray = AppsAPI.AppListResponsesV4PagePaginationArray;
})(Apps = exports.Apps || (exports.Apps = {}));
