// cloudflare@3.5.0/resources/zones/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Zones = exports.Holds = exports.Subscriptions = exports.CustomNameservers = exports.Settings = exports.ActivationCheck = void 0;
var activation_check_1 = require("./activation-check.js");
Object.defineProperty(exports, "ActivationCheck", { enumerable: true, get: function () { return activation_check_1.ActivationCheck; } });
var settings_1 = require("./settings.js");
Object.defineProperty(exports, "Settings", { enumerable: true, get: function () { return settings_1.Settings; } });
var custom_nameservers_1 = require("./custom-nameservers.js");
Object.defineProperty(exports, "CustomNameservers", { enumerable: true, get: function () { return custom_nameservers_1.CustomNameservers; } });
var subscriptions_1 = require("./subscriptions.js");
Object.defineProperty(exports, "Subscriptions", { enumerable: true, get: function () { return subscriptions_1.Subscriptions; } });
var holds_1 = require("./holds.js");
Object.defineProperty(exports, "Holds", { enumerable: true, get: function () { return holds_1.Holds; } });
var zones_1 = require("./zones.js");
Object.defineProperty(exports, "Zones", { enumerable: true, get: function () { return zones_1.Zones; } });
