// cloudflare@3.5.0/resources/addressing/prefixes/bgp/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Statuses = exports.Bindings = exports.ServiceBindingsSinglePage = exports.Prefixes = exports.BGPPrefixesSinglePage = exports.BGP = void 0;
var bgp_1 = require("./bgp.js");
Object.defineProperty(exports, "BGP", { enumerable: true, get: function () { return bgp_1.BGP; } });
var prefixes_1 = require("./prefixes.js");
Object.defineProperty(exports, "BGPPrefixesSinglePage", { enumerable: true, get: function () { return prefixes_1.BGPPrefixesSinglePage; } });
Object.defineProperty(exports, "Prefixes", { enumerable: true, get: function () { return prefixes_1.Prefixes; } });
var bindings_1 = require("./bindings.js");
Object.defineProperty(exports, "ServiceBindingsSinglePage", { enumerable: true, get: function () { return bindings_1.ServiceBindingsSinglePage; } });
Object.defineProperty(exports, "Bindings", { enumerable: true, get: function () { return bindings_1.Bindings; } });
var statuses_1 = require("./statuses.js");
Object.defineProperty(exports, "Statuses", { enumerable: true, get: function () { return statuses_1.Statuses; } });
