// cloudflare@3.5.0/resources/dns/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Settings = exports.DNS = exports.FirewallResource = exports.FirewallsV4PagePaginationArray = exports.Analytics = exports.Records = exports.RecordsV4PagePaginationArray = void 0;
var records_1 = require("./records.js");
Object.defineProperty(exports, "RecordsV4PagePaginationArray", { enumerable: true, get: function () { return records_1.RecordsV4PagePaginationArray; } });
Object.defineProperty(exports, "Records", { enumerable: true, get: function () { return records_1.Records; } });
var index_1 = require("./analytics/index.js");
Object.defineProperty(exports, "Analytics", { enumerable: true, get: function () { return index_1.Analytics; } });
var index_2 = require("./firewall/index.js");
Object.defineProperty(exports, "FirewallsV4PagePaginationArray", { enumerable: true, get: function () { return index_2.FirewallsV4PagePaginationArray; } });
Object.defineProperty(exports, "FirewallResource", { enumerable: true, get: function () { return index_2.FirewallResource; } });
var dns_1 = require("./dns.js");
Object.defineProperty(exports, "DNS", { enumerable: true, get: function () { return dns_1.DNS; } });
var settings_1 = require("./settings.js");
Object.defineProperty(exports, "Settings", { enumerable: true, get: function () { return settings_1.Settings; } });
