// cloudflare@3.5.0/resources/intel/intel.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../resource.mjs";
import * as DNSAPI from "./dns.mjs";
import * as DomainHistoryAPI from "./domain-history.mjs";
import * as IPListsAPI from "./ip-lists.mjs";
import * as IPsAPI from "./ips.mjs";
import * as MiscategorizationsAPI from "./miscategorizations.mjs";
import * as SinkholesAPI from "./sinkholes.mjs";
import * as WhoisAPI from "./whois.mjs";
import * as ASNAPI from "./asn/asn.mjs";
import * as AttackSurfaceReportAPI from "./attack-surface-report/attack-surface-report.mjs";
import * as DomainsAPI from "./domains/domains.mjs";
import * as IndicatorFeedsAPI from "./indicator-feeds/indicator-feeds.mjs";
export class Intel extends APIResource {
    constructor() {
        super(...arguments);
        this.asn = new ASNAPI.ASN(this._client);
        this.dns = new DNSAPI.DNS(this._client);
        this.domains = new DomainsAPI.Domains(this._client);
        this.domainHistory = new DomainHistoryAPI.DomainHistoryResource(this._client);
        this.ips = new IPsAPI.IPs(this._client);
        this.ipLists = new IPListsAPI.IPLists(this._client);
        this.miscategorizations = new MiscategorizationsAPI.Miscategorizations(this._client);
        this.whois = new WhoisAPI.Whois(this._client);
        this.indicatorFeeds = new IndicatorFeedsAPI.IndicatorFeeds(this._client);
        this.sinkholes = new SinkholesAPI.Sinkholes(this._client);
        this.attackSurfaceReport = new AttackSurfaceReportAPI.AttackSurfaceReport(this._client);
    }
}
(function (Intel) {
    Intel.ASN = ASNAPI.ASN;
    Intel.DNS = DNSAPI.DNS;
    Intel.DNSListResponsesV4PagePagination = DNSAPI.DNSListResponsesV4PagePagination;
    Intel.Domains = DomainsAPI.Domains;
    Intel.DomainHistoryResource = DomainHistoryAPI.DomainHistoryResource;
    Intel.IPs = IPsAPI.IPs;
    Intel.IPLists = IPListsAPI.IPLists;
    Intel.Miscategorizations = MiscategorizationsAPI.Miscategorizations;
    Intel.Whois = WhoisAPI.Whois;
    Intel.IndicatorFeeds = IndicatorFeedsAPI.IndicatorFeeds;
    Intel.IndicatorFeedListResponsesSinglePage = IndicatorFeedsAPI.IndicatorFeedListResponsesSinglePage;
    Intel.Sinkholes = SinkholesAPI.Sinkholes;
    Intel.SinkholesSinglePage = SinkholesAPI.SinkholesSinglePage;
    Intel.AttackSurfaceReport = AttackSurfaceReportAPI.AttackSurfaceReport;
})(Intel || (Intel = {}));
